Engine/Commands/InvokeScriptAnalyzerCommand.cs

```diff
@@ -402,29 +427,32 @@ protected override void StopProcessing()
 
         private void ProcessInput()
         {
-            IEnumerable<DiagnosticRecord> diagnosticsList = Enumerable.Empty<DiagnosticRecord>();
-            if (IsFileParameterSet())
+            WriteToOutput(RunAnalysis());
+        }
+
+        private IEnumerable<DiagnosticRecord> RunAnalysis()
+        {
+            if (!IsFileParameterSet())
             {
-                foreach (var p in processedPaths)
-                {
-                    if (fix)
-                    {
-                        ShouldProcess(p, $"Analyzing and fixing path with Recurse={this.recurse}");
-                        diagnosticsList = ScriptAnalyzer.Instance.AnalyzeAndFixPath(p, this.ShouldProcess, this.recurse);
-                    }
-                    else
-                    {
-                        ShouldProcess(p, $"Analyzing path with Recurse={this.recurse}");
-                        diagnosticsList = ScriptAnalyzer.Instance.AnalyzePath(p, this.ShouldProcess, this.recurse);
-                    }
-                    WriteToOutput(diagnosticsList);
-                }
+                return ScriptAnalyzer.Instance.AnalyzeScriptDefinition(scriptDefinition, out _, out _);
             }
-            else if (String.Equals(this.ParameterSetName, "ScriptDefinition", StringComparison.OrdinalIgnoreCase))
+
+            var diagnostics = new List<DiagnosticRecord>();
+            foreach (string path in this.processedPaths)
             {
-                diagnosticsList = ScriptAnalyzer.Instance.AnalyzeScriptDefinition(scriptDefinition, out _, out _);
-                WriteToOutput(diagnosticsList);
+                if (fix)
+                {
+                    ShouldProcess(path, $"Analyzing and fixing path with Recurse={this.recurse}");
```

Comment:
Is this a bug with the original code? If I read this correctly we will take the action no matter what, but `ShouldProcess` is supposed to notify the user and not take the action in the case of `-WhatIf`. this code will fix the script even with `-WhatIf`. I get that it's not as big a deal for just the diagnosis part, but the fix part seems wrong.

Reply:
Yeah I saw this and had the same thought. Definitely something we should review. If @bergmeister agrees, I can open an issue to track it so we can fix it beyond this PR.

Reply:
This goes back to my very first PR in PSSA, where I added the `-Fix` switch and since this is an action that changes state, I implemented `ShouldProcess` for it and declared that on the cmdlet. Since the `-Fix` parameter is only on the `-File` parameter set, I did not implement it for the `-ScriptDefinition` parameter set. Original commit is here: https://github.com/PowerShell/PSScriptAnalyzer/pull/817/commits/bfa1c5457ce2a96aaca44d4cde2593aee48293e8
Therefore I still think it's not really important here, very rarely does one run PSSA where the command would take ages and performing a dry-run would save time. Still happy to have a tracking issue